File: 1463-cherry-pickup-ii/1463-cherry-pickup-ii.cpp
class Solution {
public:
    
    int solve(int r, int c1, int c2, vector<vector<int>> &grid) {
        if (r == n || c1 < 0 || c1 >= m || c2 < 0 || c2 >= m) return 0;
        int &ret = dp[r][c1][c2];
        if (ret != -1) return ret;
        ret = 0;
        int cur = grid[r][c1] + grid[r][c2];
        if (c1 == c2) {
            cur -= grid[r][c1];
        }
        ret = max(ret, solve(r + 1, c1 + 1, c2 + 1, grid));
        ret = max(ret, solve(r + 1, c1 + 1, c2, grid));
        ret = max(ret, solve(r + 1, c1 + 1, c2 - 1, grid));
        
        ret = max(ret, solve(r + 1, c1, c2 + 1, grid));
        ret = max(ret, solve(r + 1, c1, c2, grid));
        ret = max(ret, solve(r + 1, c1, c2 - 1, grid));
        
        ret = max(ret, solve(r + 1, c1 - 1, c2 + 1, grid));
        ret = max(ret, solve(r + 1, c1 - 1, c2, grid));
        ret = max(ret, solve(r + 1, c1 - 1, c2 - 1, grid));
        
        ret += cur;
        
        return ret;
    }
    
    int cherryPickup(vector<vector<int>>& grid) {
        n = grid.size();
        m = grid[0].size();
        dp.resize(n, vector<vector<int>> (m, vector<int>(m, -1)));
        
        return solve(0, 0, m-1, grid);
    }
private:
    int n, m;
    vector<vector<vector<int>>> dp;
};

File: 1463-cherry-pickup-ii/1463-cherry-pickup-ii.java
class Solution {
    int n, m;
    public int cherryPickup(int[][] grid) {
        n = grid.length;
        m = grid[0].length;
        Integer[][][] dp = new Integer[n][m][m];
        
        return grid[0][0] + grid[0][m-1] + solve(0, 0, m - 1, dp, grid);
    }
    
    public int solve(int row, int col1, int col2, Integer[][][] dp, int[][] grid) {
        if (row == (n - 1)) return 0;
        if (dp[row][col1][col2] == null) {
            int ret = 0;
            for (int i = -1; i < 2; i++) {
                for (int j = -1; j < 2; j++) {
                    int f1 = col1 + i;
                    int f2 = col2 + j;
                    if (f1 < 0 || f1 >= m || f2 < 0 || f2 >= m) continue;
                    if (f1 == f2) {
                        ret = Math.max(ret, grid[row + 1][f1] + solve(row + 1, f1, f2, dp, grid));
                    } else {
                        ret = Math.max(ret, grid[row + 1][f1] + grid[row + 1][f2] + solve(row + 1, f1, f2, dp, grid));
                    }
                }
            }
            dp[row][col1][col2] = ret;
        }
        return dp[row][col1][col2];
    }
}